modules/packages/UnrolledLinkedList.chpl

```diff
@@ -0,0 +1,1245 @@
+      _commonInitFromIterable(other);
+    }
+
+    /*
+      Initializes a unrolledLinkedList containing elements that are copy initialized from
+      the elements contained in a list.
+
+      :arg other: The list to initialize from.
+    */
+    proc init=(other: unrolledLinkedList(this.type.eltType, ?p)) {
+      if !isCopyableType(this.type.eltType) then
+        compilerError("Cannot construct unrolledLinkedList from list with " +
+        "element type that cannot be copied");
+      _checkType(this.type.eltType);
+      _checkNodeCapacity(other.nodeCapacity);
+
+      this.eltType = this.type.eltType;
+      this.parSafe = this.type.parSafe;
+      this.nodeCapacity = other.nodeCapacity;
+      this.complete();
+      _commonInitFromIterable(other);
+    }
+
+    pragma "no doc"
+    proc _commonInitFromIterable(iterable) {
+      for x in iterable do
+        append(x);
+    }
+
+    /*
+    pragma "no doc"
+    proc _firstTimeInitializeArrays() {
+      _sanity(_head == nil);
+      _sanity(_tail == nil);
+      _sanity(_size == 0);
+      _head = new unmanaged _linkedNode(eltType);
+      _tail = _head;
+    }
+    */
+
+    //
+    // A helper function for getting a reference to a unrolledLinkedList element.
+    //
+    pragma "no doc"
+    inline proc const ref _getRef(idx: int) ref {
+      _sanity(idx >= 0 && idx < _size);
+      _sanity(_head != nil);
+      var cur = _head, pos = 0;
+      while cur != nil {
+        if idx < pos + cur!.size {
+          ref result = cur!.data[idx - pos];
+          return result;
+        }
+        else {
+          pos += cur!.size;
+        }
+        cur = cur!.next;
+      }
+      halt("unrolledLinkedList out of range");
+    }
+
+    pragma "no doc"
+    inline proc _enter() {
+      if parSafe then
+        _lock$.lock();
+    }
+
+    pragma "no doc"
+    inline proc _leave() {
+      if parSafe then
+        _lock$.unlock();
+    }
+
+    pragma "no doc"
+    inline proc const _withinBounds(idx: int): bool {
+      return (idx >= 0 && idx < _size);
+    }
+
+    /*
+     Split a node from the given node p
+     by moving half of the content of p into the new one
+    */
+    pragma "no doc"
+    proc _split(p: unmanaged _linkedNode(eltType)) {
+      _sanity(p.size == nodeCapacity);
+      var node = new unmanaged _linkedNode(eltType, nodeCapacity);
+      var start = p.size/2;
+      for i in start..p.size-1 {
+        node.append(p.data[i]);
+      }
+      p.size -= node.size;
+      _sanity(p.size > 0);
+
+      // link the new node to the old node
+      node.next = p.next;
+      p.next = node;
+
+      if p == _tail then _tail = p.next;
+
+      return node;
+    }
+
+    /*
+      If it's possible to merge this node and the next node, then do it
+      return if merged
+    */
+    pragma "no doc"
+    proc _merge(p: unmanaged _linkedNode(eltType)): bool {
+      var result = false;
+      if p.next == nil then return result;
+      if p.next!.size + p.size <= nodeCapacity {
```

Comment:
Did you consider applying hysteresis here?  As it stands, I think given a ULL with a single node that's completely full, the sequence `append(), pop(), append(), pop(), ...` will result in splitting and merging with every pair of operations.

Reply:
Hmm...That's a very bad case. Maybe there's something wrong with the implementation. I will get back to it soon.

Reply:
This snippet has been refactored to behave as [here](https://brilliant.org/wiki/unrolled-linked-list/) described. However, the case you mentioned still happens. But I think the point is that every operation takes O(nodeCapacity).